File: javascript/059.js
//  059
//  Spiral Matrix II
//  2015-12-16
/////////////////////////////////////////////////////
/**
 * @param {number} n
 * @return {number[][]}
 */
var generateMatrix = function(n) {
    var m = [];
    for (var i = 0; i < n; i++) {
        m[i] = new Array(n);
    }
    fill(m, 0, n-1, 0);
    return m;
};

function fill(m, p, q, count) {
    if (p == q) {
        m[p][p] = ++count;
        return;
    }
    else if (p > q) {
        return;
    }
    for (var i = p; i < q; i++) {
        m[p][i] = ++count;
    }
    for (i = p; i < q; i++) {
        m[i][q] = ++count;
    }
    for (i = q; i > p; i--) {
        m[q][i] = ++count;
    }
    for (i = q; i > p; i--) {
        m[i][p] = ++count;
    }
    fill(m, p+1, q-1, count);
}

/////////////////////////////////////////////////////
var n = 3;
var m = generateMatrix(n);
for (var i = 0; i < n; i++) {
    console.log(m[i]);
}

File: ruby/059.rb
# 059
# Spiral Matrix II
# 2015-12-16
#####################################################
# @param {Integer} n
# @return {Integer[][]}
def generate_matrix(n)
  m = Array.new(n) {Array.new(n)}
  fill(m, 0, n-1, 0)
  m
end
def fill(m, p, q, count)
  if p == q
    m[p][p] = count+=1
    return
  end
  return if p > q
  p.upto(q-1).each {|i| m[p][i] = count+=1}
  p.upto(q-1).each {|i| m[i][q] = count+=1}
  q.downto(p+1).each {|i| m[q][i] = count+=1}
  q.downto(p+1).each {|i| m[i][p] = count+=1}
  fill(m, p+1, q-1, count)
end

#####################################################

n = 3
m = generate_matrix(n)
m.each do |v|
  print v
  puts
end

